DataDog/datadog-agent
# [CWS] Add ebpfless functional tests

### What does this PR do?

It bootstraps the functional tests for ebpfless mode.

Tl;dr:
* Added ebpfless tests:
- * 30 tests are passing.
- * 36 are skipped because we lack some support.
- * 1 is failing (but should not).
* Test fixed: 6
* Ebpfless fixes: 4
* Gaps identified: 25


To do so:
* I've added an async connection mode to the ptracer (aka `cws-instrumentation trace` cmd). Now with a `--async`, it will directly start to trace the program, and wait for a client to start sending pending messages.
* module_tester was also adapted: it could globally be configured in ebpfless mode if env `EBPFLESS` env variable is set; allowing to run the already existing epbf tests in ebpfless mode (but a new config option is also avaible to build specific ebpfless tests if needed).
* Test module reloading (between tests) has been disable, only the first instance will be up. This is an ~hard limitation: if we reload the module, we'll reset the ptracer connection and lose the process cache. Depending of future tests, this could become an issue.

Functional tests fixed:
* By checking the timeout error:
- * TestOpenDiscarded
- * TestProcessContext/args-overflow-list-50
- * TestProcessContext/args-overflow-list-500
- * TestProcessContext/envs-overflow-list-50
- * TestProcessContext/args-overflow-list-500
* TestProcessPIDVariable: remove wrong error check

Ebpfless other fixes/improvements:
* Exit events were caught but not sent. Now they are. This is a first basic implementation and lacks of context (cause, exit code and timestamp)
* Fix `execveat` segfault when passing an empty path (special case that was not handled)
* Fix chained exec events no different threads (where only the last exec was sent multiple times)
* Fix create_at field wich was missing (to handle secl rules based on it)

Remaining gaps to have a full open/exec coverage:
* Add proper timestamps to exec/exit events
* Add more context to exits (cause and return code)
* Checks on lineage were skipped (the fact to always have parent/ancestors and finish on PID 1). We may want to let them skipped, or revisit more globally ebpfless process contexts.
* Same for "hostname", as we can't determine it.
* Event (exec/exit/open) schema validation were also skipped (because a lot of fields are actually missing). We could either wait to add them all, or work with another temporary ebpfless schema. TODO: ask backend if receiving ebpfless events with unvalid schema trigger errors.
* Inodes checks are skipped. We should consider providing them and add the checks back
* Support SYS_CREAT
* Support SYS_TRUNCATE
* Support open_by_handle_at
* Support io_uring (maybe impossible?)
* MTime/CTime
* uid/gid for open files events
* Add user/group
* Follow links?
* args/envs truncation
* tty
* scoped variable and event timestamp (which should had work ootb)
* fsuid/fsgid
* be able to wrote rules on setuid/setgid (should work except for fs[ug]id).
* capset rules
* is_thread
* busybox special cases
* fileless / memfd
* other easy FIM syscalls: mkdir, rename, unlink, close, ...
* kill action


TODO: add a gitlab job


### Motivation

* Validate what we already built
* Have automatic tests for future improvements

### Additional Notes

Also, while trying to make most of the open/exec tests pass in ebpfless, I encountered some tests that were broken.
I guess that if we take a closer look on all other existing tests we shall have more surprises.

### Possible Drawbacks / Trade-offs

### Describe how to test/QA your changes

First, build the cws-instrumentation tool:

> inv -e cws-instrumentation.build

Then launch the ebpf testsuite:

> inv -e security-agent.ebpfless-functional-tests --skip-linters --testflags "-test.v"


### Reviewer's Checklist


- [ ] Adequate QA/testing plan information is provided. Except if the `qa/skip-qa` label, with required either `qa/done` or `qa/no-code-change` labels, are applied.
